doc/man/gnunet-nse.1
.\" This file is part of GNUnet.
.\" Copyright (C) 2001-2019 GNUnet e.V.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the file
.\" FDL-1.3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/fdl.html}.
.\"
.\" Alternately, this document is also available under the General
.\" Public License, version 3 or later, as published by the Free Software
.\" Foundation.  A copy of the license is included in the file
.\" GPL3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/gpl.html
.\"
.\" SPDX-License-Identifier: GPL3.0-or-later OR FDL1.3-or-later
.\"
.Dd March 6, 2019
.Dt GNUNET-NSE 1
.Os
.Sh NAME
.Nm gnunet-nse
.Nd show network size estimates from NSE service
.Sh SYNOPSIS
.Nm
.Op Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
.Op Fl h | -help
.Op Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
.Op Fl l Ar LOGFILE | Fl -logfile= Ns Ar LOGFILE
.Op Fl v | -version
.Sh DESCRIPTION
.Nm
is a command line tool to show network size estimates from
the NSE service of GNUnet.
.Bl -tag -width indent
.It Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
Use the configuration file FILENAME.
.It Fl h | -help
Print short help on options.
.It Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
Use LOGLEVEL for logging.
Valid values are DEBUG, INFO, WARNING and ERROR.
.It Fl l Ar LOGFILE | Fl -logfile= Ns Ar LOGFILE
Configure logging to write logs to LOGFILE.
.It Fl m | -monitor
In monitor mode, gnunet-nse will continuously print the NSE estimates,
instead of giving just a snapshot.
.It Fl v | -version
Print GNUnet version number.
.El
.Sh FILES
.Pa gnunet.conf
Configuration file for gnunet.
.Sh SEE ALSO
.Xr gnunet.conf 5
.sp
The full documentation for gnunet is maintained as a Texinfo manual.
If the
.Xr info 1
gnunet programs are properly installed at your site, the command
.Bd -literal -offset indent -compact
        info gnunet
.Ed
should give you access to the complete handbook,
.Bd -literal -offset indent -compact
        info gnunet-c-tutorial
.Ed
will give you access to a tutorial for developers.
.Pp
Depending on your installation, this information is also
available in
.Xr gnunet 7 and
.Xr gnunet-c-tutorial 7 .
.Sh HISTORY
This man page first appeared in GNUnet 0.11.1.
.Sh AUTHORS
This page was originally written by
.An Bertrand Marc Aq Mt bmarc@debian.org
for Debian's gnunet package (man page date: 2014).
Further edits and conversion to mdoc were done by
.An ng0 Aq Mt ng0@gnunet.org .
.Sh BUGS
Report bugs by using
.Lk https://bugs.gnunet.org
or by sending electronic mail to
.Aq Mt bug-gnunet@gnu.org .


NEWS
v0.21.1:
  - Communicator and connectivity bugfixes

v0.21.0:
  - Reworked PEERSTORE API
  - Added record flag for maintenance records
  - ensure traits can be generated with subsystem-specific prefixes for the symbols
  - libgnunettesting first major testing NG refactor towards getting dependency structure streamlined
  - Remove single-use API macro GNUNET_VA_ARG_ENUM
  - major revision of blind signature API
  - Introduced closure to hold store context when caling function to add hello in peerstore.
  - Added DDLs for handling GNUNET_PEERSTORE_StoreHelloContext
  - Removed old hello functionality.
  - Refactoring components under src/ into lib/, plugin/, cli/ and service/
  - add support for encoding/decoding double values as part of JSON to libgnunetjson
  - Changed method GNUNET_HELLO_builder_get_expiration_time to not need parameter GNUNET_HELLO_Builder.
  - Code moved to the core package to get rid of circular dependencies.
  - Moved code to testing to have more generic test setup, which can be used not only from within transport.
  - The old hello design replaced by the new  hello design.
  - Added api to get notified when hellos are stored with peerstore service.
  - Added api to store hellos with peerstore service.
  - Changed new hello uri api to allow to change the expiration time
  - Moved start peer command to testing subsystem.
  - Removed all usage of old transport api, beside peerinfo tool, gnunet-transport cli and usage in transport layer itself.
  - Added __attribute__((deprecated)) to the old transport API

v0.20.0:
  - GNUNET_TESTING_get_testname_from_underscore renamed to GNUNET_STRINGS_get_suffix_from_binary_name and moved from libgnunettesting to libgnuneutil
  - Move GNUNET_s into libgnunetutil.
  - re-introduce compiler annotation for array size in signature
  - function-signature adjustment due to compiler error
  - GNUNET_PQ_get_oid removed, GNUNET_PQ_get_oid_by_name improved
  - Added GNUNET_PQ_get_oid_by_name
  - added GNUNET_PQ_get_oid()
  - Added new CCA-secure KEM and use in IDENTITY encryption
  - Add KEM API to avoid ephemeral private key management
  - Add new GNUNET_PQ_event_do_poll() API to gnunet_pq_lib.h
  - Added API to support arrays in query results
  - Improve PQ API documentation.
  - API for array types extended for times
  - API extended for array query types
  - relevant array-types in queries (not results) in postgresql added
  - just style fixes, int to enum
  - initial steps towards support of array-types in posgresql
  - adds GNUNET_JSON_spec_object_const() and GNUNET_JSON_spec_array_const()

v0.19.4:
  - No changes

v0.19.3:
  - We now detect MySQL's strange, version-dependent my_bool type on configure.
  - Add pkg-config definitions for gnunet messenger.

v0.19.2:
  - New changelog format origin


src/cli/nse/gnunet-nse.c
/*
      This file is part of GNUnet
      Copyright (C) 2008--2014, 2016 GNUnet e.V.

      GNUnet is free software: you can redistribute it and/or modify it
      under the terms of the GNU Affero General Public License as published
      by the Free Software Foundation, either version 3 of the License,
      or (at your option) any later version.

      GNUnet is distributed in the hope that it will be useful, but
      WITHOUT ANY WARRANTY; without even the implied warranty of
      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
      Affero General Public License for more details.

      You should have received a copy of the GNU Affero General Public License
      along with this program.  If not, see <http://www.gnu.org/licenses/>.

     SPDX-License-Identifier: AGPL3.0-or-later
 */

/**
 * @file nse/gnunet-nse.c
 * @brief Program to display network size estimates from the NSE service
 * @author Sree Harsha Totakura <sreeharsha@totakura.in>
 */

#include "platform.h"
#include "gnunet_nse_service.h"
#include <gnunet_util_lib.h>

/**
 * The handle to the NSE service
 */
static struct GNUNET_NSE_Handle *nse;

/**
 * The program status; 0 for success.
 */
static int status;

/**
 * Monitor flag.
 */
static int monitor;


/**
 * Task to shutdown and clean up all state
 *
 * @param cls NULL
 */
static void
do_shutdown (void *cls)
{
  (void) cls;
  if (NULL != nse)
  {
    GNUNET_NSE_disconnect (nse);
    nse = NULL;
  }
}


/**
 * Callback to call when network size estimate is updated.
 *
 * @param cls NULL
 * @param timestamp server timestamp
 * @param estimate the value of the current network size estimate
 * @param std_dev standard deviation (rounded down to nearest integer)
 *                of the size estimation values seen
 */
static void
handle_estimate (void *cls,
                 struct GNUNET_TIME_Absolute timestamp,
                 double estimate,
                 double std_dev)
{
  (void) cls;
  status = 0;
  fprintf (stdout,
           "%s: #peers=%f (ld(#peers)=%f, stddev=%f)\n",
           GNUNET_STRINGS_absolute_time_to_string (timestamp),
           GNUNET_NSE_log_estimate_to_n (estimate),
           estimate,
           std_dev);
  if (! monitor)
    GNUNET_SCHEDULER_shutdown ();
}


/**
 * Actual main function that runs the emulation.
 *
 * @param cls unused
 * @param args remaining args, unused
 * @param cfgfile name of the configuration
 * @param cfg configuration handle
 */
static void
run (void *cls,
     char *const *args,
     const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *cfg)
{
  (void) cls;
  (void) args;
  (void) cfgfile;
  nse = GNUNET_NSE_connect (cfg, &handle_estimate, NULL);
  GNUNET_SCHEDULER_add_shutdown (&do_shutdown, NULL);
}


/**
 * Main function.
 *
 * @return 0 on success
 */
int
main (int argc, char *const *argv)
{
  struct GNUNET_GETOPT_CommandLineOption options[] = {
    GNUNET_GETOPT_option_flag ('m',
                               "monitor",
                               gettext_noop (
                                 "Monitor and output current estimates"),
                               &monitor),
    GNUNET_GETOPT_OPTION_END
  };

  status = 1;
  if (GNUNET_OK !=
      GNUNET_PROGRAM_run (argc,
                          argv,
                          "gnunet-nse",
                          gettext_noop (
                            "Show network size estimates from NSE service."),
                          options,
                          &run,
                          NULL))
    return 2;
  return status;
}
